cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/JsonPaginatedListResponder.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.gateway.handlers;
+
+import com.google.gson.Gson;
+import io.cdap.http.HttpResponder;
+
+import java.io.IOException;
+import java.util.function.Function;
+
+/**
+ * JsonPaginatedListResponder - for streaming applications using Pagination
+ */
+public class JsonPaginatedListResponder extends JsonListResponder {
+
+  private String nextPageToken;
+
+  private JsonPaginatedListResponder(Gson gson, HttpResponder responder) {
+    super(gson, responder);
+
+  }
+
+  /**
+   * Factory method to be create an instance of JsonPaginatedListResponder
+   * @param gson - instance of Gson library to be used for json conversions
+   * @param responder - httpresponder to be used to stream Json objects to
+   * @param consumer - Transformation Function which will be applied the instance of JsonPaginatedListResponder
+   */
+  public static void respond(Gson gson, HttpResponder responder,
+      Function<JsonListResponder, String> consumer) throws IOException {
+    JsonPaginatedListResponder jsonPaginatedListResponder = new JsonPaginatedListResponder(gson, responder);
+      String nextPageToken = consumer.apply(jsonPaginatedListResponder);
+      jsonPaginatedListResponder.setNextPageToken(nextPageToken);
+      jsonPaginatedListResponder.finish();
+  }
+
+  @Override
+  protected void startResponse() throws IOException {
+    jsonWriter.beginObject();
+    jsonWriter.name("applications");
```

Comment:
You want this to be configurable (passed to `respond` method). This way this object can be reused for other responses